userdocs/diagnostics/expression-too-complex.md

```diff
@@ -0,0 +1,16 @@
+# Type Inference in Complex Expressions
+For some complex expressions, the Swift compiler is unable to infer types in reasonable time. This can occur for a number of reasons, but some common ones include:
+
+- The expression makes use of many overloaded operators and/or functions.
+- The expression contains literals which can be used as values of many different types (e.g. integer and floating point literals).
+- The expression includes a collection literal with a large number of nontrivial element expressions.
+
+These types of expressions can take a long time to typecheck because the compiler must consider and eliminate a larger number of possible solutions when assigning types to subexpressions.
+
+When this error occurs, possible solutions include:
+
+- Breaking the expression up into multiple subexpressions by storing intermediate values as `let` constants. These simpler subexpressions can have their types inferred more quickly.
+- Providing manual type annotations for parts of the expression. This reduces the number of types the compiler must infer, and as a result the number of possible solutions it must consider.
+- Reducing the number of overloaded declarations. New overloads of common operators like `+` or global functions like `max` may lead to slower type inference in some contexts.
+
+Sometimes, implementation limitations may cause the compiler to be unable to infer types for simpler expressions as well. If this appears to be the case, please consider filing a bug at [bugs.swift.org](http://bugs.swift.org) with an example that reproduces the issue.
```

Comment:
Perhaps some guidance as to what sorts of expressions qualify as "simpler"?

Reply:
I consider overwhelming majority of cases of `too complex` errors to be compiler bugs. It would be good to re-phrase this to ask developers to submit bug reports for any instances of such behavior, that would allow us to provide a workaround and get test-case added to the suite so we can verify against it later.

Reply:
Sounds good to me!